pubsub/api/PubsubTest.cs

```diff
@@ -0,0 +1,412 @@
+            subscriber)
+        {
+            // [START list_subscriptions]
+            string projectName = PublisherClient.FormatProjectName(_projectId);
+            IEnumerable<Subscription> subscriptions = 
+                subscriber.ListSubscriptions(projectName);
+            // [END list_subscriptions]
+            return subscriptions;
+        }
+
+        public void DeleteSubscription(string subscriptionId, SubscriberClient 
+            subscriber)
+        {
+            // [START delete_subscription]
+            string subscriptionName =
+                SubscriberClient.FormatSubscriptionName(_projectId, 
+                subscriptionId);
+            subscriber.DeleteSubscription(subscriptionName);
+            // [END delete_subscription]
+        }
+
+        public void DeleteTopic(string topicId, PublisherClient publisher)
+        {
+            // [START delete_topic]
+            string topicName = PublisherClient.FormatTopicName(_projectId, 
+                topicId);
+            publisher.DeleteTopic(topicName);
+            // [END delete_topic]
+        }
+
+        // [START retry]
+        /// <summary>
+        /// Retry the action when a Grpc.Core.RpcException is thrown.
+        /// </summary>
+        private T RetryRpc<T>(Func<T> action)
+        {
+            List<Grpc.Core.RpcException> exceptions = null;
+            var delayMs = _retryDelayMs;
+            for (int tryCount = 0; tryCount < _retryCount; ++tryCount)
+            {
+                try
+                {
+                    return action();
+                }
+                catch (Grpc.Core.RpcException e)
+                {
+                    if (exceptions == null)
+                        exceptions = new List<Grpc.Core.RpcException>();
+                    exceptions.Add(e);
+                }
+                System.Threading.Thread.Sleep(delayMs);
+                delayMs *= 2;  // Exponential back-off.
+            }
+            throw new AggregateException(exceptions);
+        }
+
+        private void RetryRpc(Action action)
+        {
+            try {
+                RetryRpc(() => { action(); return 0; });
+            }
+            catch (AggregateException exceptions)
+            {
+                throw exceptions;
+            }
+        }
+
+        public void RpcRetry(string topicId, string subscriptionId,
+            PublisherClient publisher, SubscriberClient subscriber)
+        {
+            string topicName = PublisherClient.FormatTopicName(_projectId, topicId);
+            string subscriptionName =
+                SubscriberClient.FormatSubscriptionName(_projectId, subscriptionId);
+            try
+            {
+                RetryRpc(() =>
+                {
+                    // Subscribe to Topic
+                    // This should fail since Topic has not yet been created
+                    subscriber.CreateSubscription(subscriptionName, topicName,

```

Comment:
I see why we want to have this code in the test, so that we can test the retry behavior and make sure an exception gets appropriately thrown.

However, I'm not sure this code should live between the // [START retry] and // [END retry] tags.